#  Armadillo Data Complete Analysis
In this notebook we will find all the information from loading the data and analysing it. 
Note that for running it in your computer you will need to use the command "read.table" after you set your working directory. 


In [ ]:
library(tidyverse)

arm<-tbl_df(read.table("armadillo2.txt",header=T)) %>%
  mutate(farm=as.factor(farm),sex=as.factor(sex))
arm

In the next cell we will have a similar set of commands to calculate farm, sex and snp groups means:

In [7]:
#calculating farm average:
farm_mean <- arm %>% group_by(farm) %>% summarize(X_farm=mean(weight))
farm_mean #when we execute this line it is the same thing as typing its name in the console

#now sex average:
sex_mean <- arm %>% group_by(sex) %>% summarize(X_sex=mean(weight))
sex_mean

#and for snps #I am not saving SNPs means because this type of analysis is not correct :)
arm %>% group_by(snp1) %>% summarize(SNP1_mu=mean(weight))
arm %>% group_by(snp2) %>% summarize(SNP2_mu=mean(weight))
arm %>% group_by(snp3) %>% summarize(SNP3_mu=mean(weight))

#for the last SNP I will show how we can separate pipes in different lines
arm %>%  #you can include comments after each pipe
  group_by(snp4) %>%  #note that the pipe has to come at the end of the line
  summarize(SNP4_mu=mean(weight))



farm X_farm  
1 1    117.2878
2 2    108.3620
3 3    111.7864
4 4    113.1971

sex X_sex   
1 1   110.1092
2 2   115.5799

snp1 SNP1_mu 
1 0    114.6230
2 1    112.0070
3 2    112.6415

snp2 SNP2_mu 
1 0    113.7025
2 1    113.4119
3 2    112.1385

snp3 SNP3_mu 
1 0    111.6357
2 1    113.2715
3 2    114.3285

snp4 SNP4_mu 
1 0    111.0065
2 1    114.2121
3 2    114.3973

The regression we will try as we did before. We will start with a very simple model, and then add complexity. We need to be 100% sure that we are treating the sex and farm effect as factors, and the snps as numeric, otherwise we will have problems in our regressions.

In [8]:

model1<-lm(weight~farm,arm) #make sure you have farm as a factor
summary(model1) #more information about out regression
#are the estimates similar to the means we've calculated earlier?



Call:
lm(formula = weight ~ farm, data = arm)

Residuals:
     Min       1Q   Median       3Q      Max 
-13.7170  -2.9402  -0.1569   3.0608  15.3236 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 117.2878     0.6385 183.705  < 2e-16 ***
farm2        -8.9257     1.0049  -8.882 4.20e-16 ***
farm3        -5.5013     0.9982  -5.511 1.11e-07 ***
farm4        -4.0907     0.9918  -4.125 5.48e-05 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 5.147 on 196 degrees of freedom
Multiple R-squared:  0.2981,	Adjusted R-squared:  0.2873 
F-statistic: 27.74 on 3 and 196 DF,  p-value: 5.349e-15


In [9]:
#let's define a more complex model
#Weight = farm + sex 

model2<-lm(weight~farm+sex,arm) #make sure you have farm and sex as a factor
summary(model2) #more information about out regression
#now we will see that the intercept is farm=1 AND sex=1. If we have farm=1 and sex=2, we would add 5.48
#if we have farm=2 and sex=1 we would add -9.00; if we have farm=2 and sex=2 we would add -9.00 + 5.48


Call:
lm(formula = weight ~ farm + sex, data = arm)

Residuals:
     Min       1Q   Median       3Q      Max 
-14.2324  -2.8100   0.0526   2.6811  13.1769 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 114.2493     0.6422 177.908  < 2e-16 ***
farm2        -9.0044     0.8523 -10.564  < 2e-16 ***
farm3        -5.1450     0.8476  -6.070 6.55e-09 ***
farm4        -4.3916     0.8419  -5.217 4.64e-07 ***
sex2          5.4861     0.6233   8.802 7.23e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.366 on 195 degrees of freedom
Multiple R-squared:  0.4977,	Adjusted R-squared:  0.4873 
F-statistic: 48.29 on 4 and 195 DF,  p-value: < 2.2e-16


In [10]:
#one more effect :) #Weight = farm + sex + snp1
model3<-lm(weight~farm+sex+snp1,arm) #make sure you have farm and sex as a factor, but snp as numeric (dbl)
summary(model3) #more information about out regression
#we have a single value for snp1. The estimate of the snp effect will be multiplied by the "gene content"
#ex: SNP1=0: we add nothing; SNP1=1 we add 1*(-0.78), and SNP1=2 we add 2*(-0.78)


Call:
lm(formula = weight ~ farm + sex + snp1, data = arm)

Residuals:
     Min       1Q   Median       3Q      Max 
-14.2284  -2.5990  -0.2199   2.6864  13.3161 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 115.0264     0.7374 155.983  < 2e-16 ***
farm2        -8.9496     0.8455 -10.585  < 2e-16 ***
farm3        -5.1295     0.8404  -6.104 5.54e-09 ***
farm4        -4.4721     0.8356  -5.352 2.44e-07 ***
sex2          5.4356     0.6185   8.789 8.05e-16 ***
snp1         -0.7854     0.3760  -2.089    0.038 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 4.329 on 194 degrees of freedom
Multiple R-squared:  0.5087,	Adjusted R-squared:  0.496 
F-statistic: 40.17 on 5 and 194 DF,  p-value: < 2.2e-16


In [12]:

#now let's include everything in the model :) yay!
#one more effect :) #Weight = farm + sex + snp1
model4<-lm(weight~farm+sex+snp1+snp2+snp3+snp4,arm) #make sure you have farm and sex as a factor, but snp as numeric (dbl)
summary(model4) #more information about out regression
#the numbers I used over here are from armadillo 1, 
#the means and effects (specially for farms) are different in the second data. 
 


Call:
lm(formula = weight ~ farm + sex + snp1 + snp2 + snp3 + snp4, 
    data = arm)

Residuals:
     Min       1Q   Median       3Q      Max 
-11.4874  -2.2328  -0.0574   2.3127  11.2651 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 113.9987     0.9354 121.868  < 2e-16 ***
farm2        -9.4982     0.7594 -12.508  < 2e-16 ***
farm3        -5.4226     0.7401  -7.327 6.45e-12 ***
farm4        -4.5535     0.7346  -6.199 3.43e-09 ***
sex2          5.4322     0.5449   9.969  < 2e-16 ***
snp1         -0.8924     0.3322  -2.687  0.00785 ** 
snp2         -1.4643     0.3625  -4.040 7.75e-05 ***
snp3          0.9372     0.3431   2.732  0.00689 ** 
snp4          1.7259     0.3142   5.492 1.25e-07 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.805 on 191 degrees of freedom
Multiple R-squared:  0.6263,	Adjusted R-squared:  0.6106 
F-statistic: 40.01 on 8 and 191 DF,  p-value: < 2.2e-16


Let's say we have an animal with farm=1,sex=1, and all SNPs=0, we will predict its weight as:
weight=Intercept, which is 113.99
A second animal  farm=1,sex=2, and snp1=1,snp2=2,snp3=0 and snp4=1, we have it's weight predicted as:
weight=Intercept+sex2+snp1*1+snp2*2+snp3*0+snp4*1; which will be
weight=114+5.4+(-0.89)*1+(-1.46)*2+(0.93)*0+(1.72)*1
Basically, we plug the numbers from the table above. Now, let's do it for 200 animals.
There is a problem: I don't want to calculate it 200 times. Do you?
Good news: we can tell the computer to do it for us:
The command name is **predict**, and we will use it in our **model4**.


In [16]:
pred1<-predict(model4)
#It will calculate the predicted weight for all the 200 animals in our population.
#And we can include it in our data with:
arm_new<- arm %>% mutate(pred=pred1) #it will use mutate to create a new column containing our predicted weights


The quality of our prediction will be given by the correlation between the predicted and observed weight.
The closer the value is to 1, the higher our **accuracy** will be. 
                      

In [17]:
arm_new %>% summarise(correlation=cor(weight,pred))

correlation
1 0.7913903

Why do I want to predict the weight, if I already have the animal's phenotype?
The reason we are doing this regression is to learn the effects of our components (sex, farm, and SNPs) to predict phenotypes of animal we don't have the weight. As our baby armadillo data.

In [21]:

arm_baby<-tbl_df(read.table("armadillo_young.txt",header=T)) %>%
  mutate(farm=as.factor(farm),sex=as.factor(sex)) #here we are making sure our new data is using farma dn sex as factors
pred2<-predict(model4,newdata = arm_baby) #with the newdata argument we can predict weights of out of sample data
baby_arm_new<- arm_baby %>% mutate(pred=pred2) #here we are joining our predicted data to the baby armadillo data


Let's say the baby armadillo's are now adults, and we want to know if we did a good job or not when we predicted their weight back when they were babies. We can calculate the correlation between the observed weight and the predicted weight. Note that we can't know this correlation until we have the data collected. But we can have the prediction as soon as we know the coefficients.

In [22]:
baby_arm_new %>% summarise(correlation=cor(weight,pred))

correlation
1 0.7099818